accounts/abi/event.go

```diff
@@ -57,55 +57,57 @@ func (e Event) tupleUnpack(v interface{}, output []byte) error {
 	}
 
 	var (
-		value = valueOf.Elem()
-		typ   = value.Type()
+		value    = valueOf.Elem()
+		typ      = value.Type()
+		isStruct = true
 	)
-
-	if value.Kind() != reflect.Struct {
-		return fmt.Errorf("abi: cannot unmarshal tuple in to %v", typ)
+	switch value.Kind() {
+	case reflect.Struct:
+	case reflect.Slice, reflect.Array:
+		isStruct = false
+		minLen := countNonIndexedArguments(e.Inputs)
+		if value.Len() < minLen {
+			return fmt.Errorf("abi: insufficient number of elements in the list/array for unpack, want %d, got %d",
+				minLen, value.Len())
+		}
+	default:
+		return fmt.Errorf("abi: cannot unmarshal tuple into %v", typ)
 	}
 
-	j := 0
-	for i := 0; i < len(e.Inputs); i++ {
-		input := e.Inputs[i]
+	i, j := -1, 0
+	for _, input := range e.Inputs {
 		if input.Indexed {
-			// can't read, continue
+			// Indexed arguments are not packed into data
 			continue
 		} else if input.Type.T == ArrayTy {
 			// need to move this up because they read sequentially
 			j += input.Type.Size
 		}
+		i++
 		marshalledValue, err := toGoType((i+j)*32, input.Type, output)
 		if err != nil {
 			return err
 		}
 		reflectValue := reflect.ValueOf(marshalledValue)
 
-		switch value.Kind() {
-		case reflect.Struct:
+		if isStruct {
```

Comment:
as explained below - we should check the type before unpacking the arguments in order to validate the array size once and for all. We can do the switch here, but since we already did it it's easier to reuse what we have learned.